prov/util/src/util_mr.c

```diff
@@ -0,0 +1,224 @@
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#include <config.h>
+#include <stdlib.h>
+#include <fi_enosys.h>
+#include <fi_util.h>
+#include <assert.h>
+
+/* deep copy: make seperate copy of mr_attr and use context for prov_mr */
+static struct fi_mr_attr * create_mr_attr_copy(
+                        const struct fi_mr_attr *in_attr, void * prov_mr)
+{
+    struct fi_mr_attr * item;
+    struct iovec *mr_iov;
+    int i = 0;
+
+    assert(prov_mr && in_attr);
+
+    item = malloc(sizeof(struct fi_mr_attr));
+    if(!item)
+        return NULL;
+
+    *item = *in_attr;
+    item->context = prov_mr;
+    mr_iov = malloc(sizeof(struct iovec) * in_attr->iov_count);
+    if(!mr_iov)
+        return NULL;
+
+    for(i = 0; i < in_attr->iov_count; i++)
+        mr_iov[i] = in_attr->mr_iov[i];
+
+    item->mr_iov = mr_iov;
+
+    return item;
+}
+
+static uint64_t get_mr_key(struct ofi_util_mr *mr_h)
+{
+    assert(mr_h->b_key != UINT64_MAX);
+    return mr_h->b_key++;
+}
+
+static int verify_addr(struct fi_mr_attr * item, uint64_t in_access,
+                                 uint64_t in_addr, ssize_t in_len)
+{
+    int i = 0;
+    uint64_t start = (uint64_t)item->mr_iov[i].iov_base;
+    uint64_t end = start + item->mr_iov[i].iov_len;
+
+    if(!in_addr)
+        return -FI_EINVAL;
+
+    if((in_access & item->access) != in_access)
+        return -FI_EACCES;
+
+      for (i = 0; i < item->iov_count; i++) {
+        if (start <= in_addr && end >= (in_addr + in_len))
+            return 0;
+    }
+
+    return -FI_EACCES;
+}
+
+int ofi_mr_insert(struct ofi_util_mr * in_mr_h, const struct fi_mr_attr *in_attr,
+                                uint64_t * out_key, void * in_prov_mr)
+{
+    struct fi_mr_attr * item;
+
+    if (!in_attr || in_attr->iov_count <= 0 || !in_prov_mr) {
+        return -FI_EINVAL;
+    }
+
+    item = create_mr_attr_copy(in_attr, in_prov_mr);
+    if(!item)
+        return -FI_ENOMEM;
+
+    /* Scalable MR handling: use requested key and offset */
+    if(in_mr_h->mr_type == FI_MR_SCALABLE) {
+        item->offset = (uintptr_t) in_attr->mr_iov[0].iov_base + in_attr->offset;
+        /* verify key doesn't already exist */
+        if(rbtFind(in_mr_h->map_handle, &item->requested_key))
+                return -FI_EINVAL;
+    } else {
+        item->requested_key = get_mr_key(in_mr_h);
+        item->offset = (uintptr_t) in_attr->mr_iov[0].iov_base;
+    }
+
+    rbtInsert(in_mr_h->map_handle, &item->requested_key, item);
+    *out_key = item->requested_key;
+
+    return 0;
+}
+
+/* io_addr is address of buff (&buf) */
+int ofi_mr_retrieve(struct ofi_util_mr * in_mr_h, ssize_t in_len,
+                                uintptr_t *io_addr, uint64_t in_key,
+                                uint64_t in_access, void **out_prov_mr)
+{
+    /*grab info */
+    void * itr;
+    struct fi_mr_attr * item;
+    int err = 0;
+
+    assert(io_addr);
```

Comment:
probably better check with "if" against input parameters. assert only works for debug build.